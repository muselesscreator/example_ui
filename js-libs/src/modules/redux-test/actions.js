/* globals it, expect */

import _ from 'lodash';

// Action Tests
export const testActions = (actions, fnKeys) => {
  test('all action types are unique', () => {
    const types = Object.keys(actions).map(action => actions[action].type);
    expect(_.uniq(types)).toEqual(types);
    expect(_.every(types, (type) => typeof type === 'string')).toEqual(true);
  });
  Object.keys(fnKeys).forEach(actionKey => {
    describe(actionKey, () => {
      it('forwards the correct arguments to the action', () => {
        const fn = actions[actionKey].fn,
              props = fnKeys[actionKey];
        const expected = props.reduce((obj, prop) => ({ ...obj, [prop]: prop }), {})
        expect(fn(...props)).toEqual(expected);
      });
    });
  });
};

export default testActions;
