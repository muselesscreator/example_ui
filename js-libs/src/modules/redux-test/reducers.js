import _ from 'lodash';

import PropTypes from 'prop-types';

/**
 * @callback reducerFunc
 * @param {object} state - current redux store state
 * @param {object} action - redux action being handled
 * @return {object} outgoing state.
 */

/**
 * @function
 * Define a test suite for a redux reducer.
 *
 * @param {func} reducer - reducer function to test
 * @param {object} initialState - initial redux store state
 * @param {object[]} reducerTests - array of "tests" to run, each containing
 *   an action definition, a list of action args, an optional label string,
 *   and an initial and expected state.
 */
export const testReducer = (reducer, initialState, reducerTests) => {
  it('loads initial state', () => {
    expect(reducer(undefined, {})).toEqual(initialState);
  });
  reducerTests.forEach(test => {
    const { action, args=[], state, expected, label='' } = test;
    let testStr = `reacts to ${test.action.type} action type`;
    if (label !== '') {
      testStr = `${testStr}: ${label}`;
    }
    it(testStr, () => {
      const { type, fn } = action;
      const out = fn(...args);
      expect(reducer(state, { ...out, type })).toEqual(expected);
    });
  });
};
testReducer.propTypes = ({
  reducer: PropTypes.func.isRequired,
  initialState: PropTypes.func.isRequired,
  tests: PropTypes.arrayOf({
    action: PropTypes.shape({
      type: PropTypes.string.isRequired,
      fn: PropTypes.func.isRequired,
    }).isRequired,
    args: PropTypes.arrayOf(PropTypes.any).isRequired,
    expected: PropTypes.object.isRequired,
    label: PropTypes.string,
    state: PropTypes.any.isRequired,
  }).isRequired,
});

const obj = { an: 'object' };
const initial = { a: 'state', and: 'stuff' };
export const tests = {
  /**
   * @function
   * Test a partial update reducer function.
   * Provide the tested action, a target path, a set of action args,
   * and an expected merge value to be added to an object at the
   * given path.
   *
   * @param {object} action - { type, fn } action definition
   * @param {string} path - _.set-able path string ('field', 'field.group[0]', etc)
   * @param {array} args - list of action arguments
   * @param {object} expectedMerge - expected data to merge into the given field.
   * @param {string=} label - optional test label string
   * @return {object} - test object suitable for testReducer usage
   */
  objPartial: (action, path, args, expectedMerge, label) => {
    const state = { ...initial };
    const expected = { ...initial };
    _.set(state, path, obj);
    _.set(expected, path, { ...obj, ...expectedMerge });
    return { action, args, state, expected, label };
  },
};

export default testReducer;
