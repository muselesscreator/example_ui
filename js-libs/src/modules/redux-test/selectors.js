/* globals it, expect */

import PropTypes from 'prop-types';

export const listData = ['a thing', 'another thing'];

// Tests simpleSelectors.
export const testSelectors = (selectors, keys) => {
  keys.forEach(key => {
    const selector = selectors[key];
    it(`returns the ${key} from the passed state`, () => {
      expect(selector({ test: 'asdf', [key]: listData })).toEqual(listData);
    });
  });
};
testSelectors.propTypes = ({
  selector: PropTypes.func.isRequired,
  prop: PropTypes.string.isRequired
});

export default testSelectors;
