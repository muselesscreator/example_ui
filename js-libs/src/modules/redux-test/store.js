/* globals it, expect */

import PropTypes from 'prop-types';

import _ from 'lodash';

export const listData = ['a thing', 'another thing'];

// Action Tests
export const testActions = (actions, fnKeys) => {
  test('all action types are unique', () => {
    const types = Object.keys(actions).map(action => actions[action].type);
    expect(_.uniq(types)).toEqual(types);
    expect(_.every(types, (type) => typeof type === 'string')).toEqual(true);
  });
  Object.keys(fnKeys).forEach(actionKey => {
    describe(actionKey, () => {
      it('forwards the correct arguments to the action', () => {
        const fn = actions[actionKey].fn,
              props = fnKeys[actionKey];
        const expected = props.reduce((obj, prop) => ({ ...obj, [prop]: prop }), {})
        expect(fn(...props)).toEqual(expected);
      });
    });
  });
};

// Reducers 
export const testReducer = (reducer, initialState, tests) => {
  it('loads initial state', () => {
    expect(reducer(undefined, {})).toEqual(initialState);
  });
  tests.forEach(test => {
    it(`reacts to ${test.action.type} action type`, () => {
      const { action, args, state, expected } = test;
      const { type, fn } = action;
      expect(reducer(state, { ...fn(...args), type })).toEqual(expected);
    });
  });
};
testReducer.propTypes = ({
  reducer: PropTypes.func.isRequired,
  initialState: PropTypes.func.isRequired,
  tests: PropTypes.shape({
    action: PropTypes.shape({
      type: PropTypes.string.isRequired,
      fn: PropTypes.func.isRequired,
    }),
    args: PropTypes.arrayOf(PropTypes.any),
    state: PropTypes.any,
    expected: PropTypes.object,
  }).isRequired,
});

// Selector Tests
export const testSelector = (selector, prop) => {
  it(`returns the ${prop} from the passed state`, () => {
    expect(selector({ test: 'asdf', [prop]: listData })).toEqual(listData);
  });
};
testSelector.propTypes = ({
  selector: PropTypes.func.isRequired,
  prop: PropTypes.string.isRequired
});
