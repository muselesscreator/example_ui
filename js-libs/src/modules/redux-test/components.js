/* globals it, expect */

import React from 'react';
import PropTypes from 'prop-types';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import * as _ from 'lodash';

import { defaultTo } from 'modules/common';

// Components
export const render = (Component, props) => {
  return shallow(<Component {...props} />);
};
render.propTypes = ({
  Component: PropTypes.element.isRequired,
  props: PropTypes.objectOf(PropTypes.node).isRequired
});

export const snapshotTest = tree => {
  expect(toJson(tree)).toMatchSnapshot();
};

export const simpleTest = (name, Component, props) => {
  test('snapshot' + name ? `:${name}` : '' , () => {
    snapshotTest(render(Component, props));
  });
}
simpleTest.propTypes = ({
  name: PropTypes.string.isRequired,
  Component: PropTypes.node.isRequired,
  props: PropTypes.object,
});

// mapStateToProps
export const mockSelectors = (selectors, toMock) => {
  Object.keys(toMock).forEach(key => {
    _.set(
      selectors,
      toMock[key],
      jest.fn((...props) => ([ ...props ]))
    );
  });
}
mockSelectors.propTypes = ({
  selectors: PropTypes.object.isRequired,
  toMock: PropTypes.objectOf(PropTypes.string),
});

export const testSelector = (selectorProp, args) => {
  expect(selectorProp).toEqual({ ...args });
};

export const testSelectors = (mapStateToProps, ownProps, selectors, pathMap, argsMap) => {
  mockSelectors(selectors, pathMap);
  const state = 'stATe';
  const props = mapStateToProps(state, ownProps);
  Object.keys(argsMap).forEach(prop => {
    expect(props[prop]).toEqual(state, ...argsMap[prop])
  });
}

/**
 * @callback mapStateToProps
 * @param {object} state - current redux store state
 */

/**
 * @function
 * Tests simple selector-prop connections in a mapStateToProps function.
 *
 * Takes a mapStateToProps connect method, along with the base selectors collection
 * from the redux store, and an object containing a mapping of used selectors, keyed
 * by their returned prop name.
 *
 * @param {mapStateToProps} mapStateToProps - component mapStateToProps function.
 * @param {object} selectors - base redux store.selectors
 * @param {object} pathMap - { <prop name>: <selector path> }
 * @param {object=} argsMap - { <prop name>: [<selector args>] }
 * @param {object=} ownProps - optional ownProps object
 */
export const testStateToProps = (mapStateToProps, selectors, pathMap, argsMap={}, ownProps={}) => {
  mockSelectors(selectors, pathMap);
  const state = { field: 'FieLD' };
  const props = mapStateToProps(state, ownProps);
  Object.keys(pathMap).forEach(key => {
    if (argsMap[key] === undefined) {
      expect(props[key]).toEqual([state]);
    }
    else {
      expect(props[key]).toEqual([state, ...argsMap[key]])
    }
  });
};
testStateToProps.propTypes = ({
  mapStateToProps: PropTypes.func.isRequired,
  selectors: PropTypes.object.isRequired,
  pathMap: PropTypes.object.isRequired,
  argsMap: PropTypes.object,
  ownProps: PropTypes.object,
});

// mapDispatchToProps
export const mockActions = (actions, toMock) => {
  toMock.map(actionPath => _.set(actions, actionPath, jest.fn((...props) => [...props])));
}
mockActions.propTypes = {
  actions: PropTypes.object,
  toMock: PropTypes.arrayOf(PropTypes.string),
};
