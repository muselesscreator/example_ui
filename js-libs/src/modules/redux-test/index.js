import testActions from './actions';
import * as components from './components';
import testReducer, { tests as reducerTests } from './reducers';
import testSelectors from './selectors';

export {
  testActions,
  components,
  testReducer,
  reducerTests,
  testSelectors,
}
