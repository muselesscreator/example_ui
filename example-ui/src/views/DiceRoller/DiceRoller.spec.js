import { components } from 'js-libs/build/redux-test';

import * as module from './DiceRoller';

jest.mock('./components', () => ({
  DieControl: 'DieControl',
  RollsList: 'RollsList',
}));

describe('DiceRoller component', () => {
  describe('Component', () => {
    components.simpleTest('smoke test', module.DiceRoller, {});
  });
});

