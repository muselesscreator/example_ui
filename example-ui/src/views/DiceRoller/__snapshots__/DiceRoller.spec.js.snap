// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiceRoller component Component :smoke test 1`] = `
<div
  className="dice-roller-view"
>
  <div
    className="dice-container"
  >
    <DieControl
      type="d4"
    />
    <DieControl
      type="d6"
    />
    <DieControl
      type="d8"
    />
    <DieControl
      type="d10"
    />
    <br />
    <DieControl
      type="d12"
    />
    <DieControl
      type="d20"
    />
    <DieControl
      type="d100"
    />
  </div>
  <RollsList />
</div>
`;
